03week/ticTacToe.js

```diff
@@ -23,27 +31,151 @@ function printBoard() {
   console.log('2 ' + board[2].join(' | '));
 }
 
+// checks for horizontal win
+// returns true/false
+// if true, sets global whichPlayerWin var
 function horizontalWin() {
-  // Your code here
+  let isWin = false; // value to be returned
+
+  // iterate over each row in board
+  board.forEach((row) => {
+    if ((row[0] === 'X' && row[1] === 'X' && row[2] === 'X') || (row[0] === 'O' && row[1] === 'O' && row[2] === 'O')) {
+      isWin = true;
+      whichPlayerWin = (row[0] === 'X') ? 'X' : 'O'; // set which player won
+    }
+  });
+
+  return isWin;
 }
 
+// checks for vertical win
+// returns true/false
+// if true, sets global whichPlayerWin var
 function verticalWin() {
-  // Your code here
+  let isWin = false; // value to be returned
+
+  // iterate over each col in board
+  // wasnt sure how to do this in a forEach loop
+  for (let i = 0; i <= 2; i++) {
+    if ((board[0][i] === 'X' && board[1][i] === 'X' && board[2][i] === 'X') || (board[0][i] === 'O' && board[1][i] === 'O' && board[2][i] === 'O')) {
+      isWin = true;
+      whichPlayerWin = (board[0][i] === 'X') ? 'X' : 'O'; // set which player won
+      break;
+    }
+  }
+
+  return isWin;
 }
 
+// checks for diagonal win
+// returns true/false
+// if true, sets global whichPlayerWin var
 function diagonalWin() {
-  // Your code here
+  let isWin = false; // value to be returned
+
+  // if player X has diagonal win (both directions)
+  if ((board[0][0] === 'X' && board[1][1] === 'X' && board[2][2] === 'X') || (board[0][2] === 'X' && board[1][1] === 'X' && board[2][0] === 'X')) {
+    isWin = true;
+    whichPlayerWin = 'X'; // set which player won
+  } // if player O has diagonal win (both directions)
+  else if ((board[0][0] === 'O' && board[1][1] === 'O' && board[2][2] === 'O') || (board[0][2] === 'O' && board[1][1] === 'O' && board[2][0] === 'O')) {
+    isWin = true;
+    whichPlayerWin = 'O'; // set which player won
+  }
+
+  return isWin;
 }
 
+// checks for: horizontal, vertical, or diagonal win
+// returns true/false
 function checkForWin() {
-  // Your code here
+  return horizontalWin() || verticalWin() || diagonalWin();
+}
+
+// checks to make sure user input is valid
+// returns true/false
+function isValidInput(row, col) {
+  return ((row >= 0 && row <= 2) && (col >= 0 && col <= 2));
+}
+
+// updates the board with user input
+// returns true/false
+// true: if location is not occupied, updates board
+// false: if location is empty, does not update board
+function updateBoard(row, col) {
+  // check if location user chose is empty
+  if (board[row][col] === ' ') {
+    board[row][col] = playerTurn;
+    return true;
+  } // if location user chose is occupied
+  else {
```

Comment:
this else is redundant, you can erase

Reply:
ok have removed the else clause

I now understand that if a function does not explicitly return anything, it returns `undefined` which is a falsy value